### PR TITLE
Implement `size` attribute for <font> element

### PR DESCRIPTION
**Pull request by frewsxcv**
[Review on Reviewable](https://reviewable.io/reviews/servo/servo/7683)

### CODE REVIEW

**Comment by frewsxcv** [comment]
There is no WPT test for this at the time of writing (https://github.com/w3c/web-platform-tests/issues/2167). I started writing one though:

```html
<!DOCTYPE html>
<meta charset=utf-8>
<title>'font' element, 'size' attribute</title>
<link rel=help href="SPEC LINK HERE">
<link rel=author title="Corey Farwell" href="mailto:coreyf@rwell.org">
<script src="/resources/testharness.js"></script>
<script src="/resources/testharnessreport.js"></script>
<div id="log"></div>

<font>Hello</font>

<script>
var elem = document.getElementsByTagName('font')[0];

var modes = ["", "+", "-"];
var sizes = ["0", "1", "2", "3", "4", "5", "6", "7", "8"];

var computedSizes = [
  //    0,      1,      2,      3,      4,      5,      6,      7,      8
  ["10px", "10px", "13px", "16px", "18px", "24px", "32px", "48px", "48px"],  // mode: absolute
  ["16px", "18px", "24px", "32px", "48px", "48px", "48px", "48px", "48px"],  // mode: relative plus
  ["16px", "13px", "10px", "10px", "10px", "10px", "10px", "10px", "10px"],  // mode: relative minus
];

var testSize = function (attrValue, computedSize) {
  elem.setAttribute("size", attrValue);
  assert_equals(window.getComputedStyle(elem).fontSize, computedSize);
}

var args = [];
for (var i = 0; i < modes.length; i++) {
  for (var j = 0; j < sizes.length; j++) {
    var sizeAttrValue = modes[i] + sizes[j];
    args.push([
      "size=" + sizeAttrValue,
      sizeAttrValue,
      computedSizes[i][j]
    ]);
  }
}

generate_tests(testSize, args);
</script>
```

The test above passes with Gecko and Blink. It does not pass with this pull request. From what I understand Gecko and Blink **do not** follow the spec verbatim. I'm also not entirely sure if these are truly enforced by the spec or just guidelines. If anyone has suggestions, please let me know.

/cc @Ms2ger since I mentioned this a couple days ago on IRC

**Comment by nox** [comment]
@frewsxcv Is there a reftest for the content attribute?

**Comment by frewsxcv** [comment]
> Is there a reftest for the content attribute?

The test expectations for `tests/wpt/metadata/html/dom/reflection-obsolete.html` changed, so I assume so

**Comment by nox** [comment]
Oh right, I didn't notice.

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes +S-awaiting-answer 

A nit and a question.

---

Reviewed 9 of 9 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-JzimfsQkWIg_BtXgWos)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/script/dom/htmlfontelement.rs#L28)):</sup>
@jdm @Ms2ger Would you find it strange to add a SpecifiedLength constructor to AttrValue, for such presentational hints?

I don't like the idea of storing DOMString here. You could at least store an Atom (they are only 7 small strings).

---

<sup>**[components/style/properties.mako.rs, line 1925 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-Jzin7xZULduWMkngCwX)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/style/properties.mako.rs#L1925)):</sup>
Shouldn't this be moved out of here?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/style/properties.mako.rs, line 1925 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-Jzin7xZULduWMkngCwX)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/style/properties.mako.rs#L1925)):</sup>
Have any suggestions where it should live?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by nox** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

<sup>**[components/style/properties.mako.rs, line 1925 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-Jzin7xZULduWMkngCwX)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/style/properties.mako.rs#L1925)):</sup>
In values, with the rest of the constants and the function you moved. What do you think?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by jdm** [comment]
A new AttrValue variant sounds ok to me, but I'd also like to hear from @Ms2ger.

**Comment by frewsxcv** [comment]
<sup>**[components/style/properties.mako.rs, line 1925 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-Jzin7xZULduWMkngCwX)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/style/properties.mako.rs#L1925)):</sup>
Just letting you know in case you don't already: this constant is used two lines below this. I'm still fine with moving it, though `styles::values::MEDIUM_PX` seems a little awkward. Maybe something like `styles::values::FONT_MEDIUM_PX`

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by Ms2ger** [comment]
Please do use a new `AttrValue` variant.

File: components/script/dom/attr.rs [hunk 1]
```diff
@@ -86,6 +86,13 @@ impl AttrValue {
         }
     }
 
+    pub fn as_str(&self) -> &str {
```
Comment (Manishearth) [inline components/script/dom/attr.rs]: Leave a doc comment mentioning that this is panicky

Comment (frewsxcv) [inline components/script/dom/attr.rs]: :+1:

Comment (frewsxcv) [inline components/script/dom/attr.rs]: https://github.com/frewsxcv/servo/commit/20581bd94d70b99ca33a8d7507d80944b068b639

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-JzimfsQkWIg_BtXgWos)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/script/dom/htmlfontelement.rs#L28)):</sup>
Just to confirm, you're talking about creating a new enum variant for `AttrValue` right? And considering there are multiple `SpecifiedLength` structs, were you thinking this variant would be generic using some trait?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by nox** [comment]
-S-awaiting-answer 

---

Review status: 8 of 9 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 0 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-JzimfsQkWIg_BtXgWos)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/script/dom/htmlfontelement.rs#L0)):</sup>
values::specified::Length

---

<sup>**[components/style/properties.mako.rs, line 1925 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-Jzin7xZULduWMkngCwX)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/style/properties.mako.rs#L1925)):</sup>
I'm fine with FONT_MEDIUM_PX.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r1\]](https://reviewable.io:443/reviews/servo/servo/7683#-JzimfsQkWIg_BtXgWos)** ([raw file](https://github.com/servo/servo/blob/358f4fc388951adaa76ad70ff7cad8cb8dfa667d/components/script/dom/htmlfontelement.rs#L28)):</sup>
One unfortunate side effect this change: We would probably have to get rid of [this `Deref`](https://github.com/servo/servo/blob/a7208869f2903e36f9b2f540b55b50283d7df466/components/script/dom/attr.rs#L101-L112) since the underlying representation of a `specified::Length` is not a `str`. We might have to make it `as_cow_str`? Any other ideas? I might be overlooking something here...

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by jdm** [comment]
Don't we want the variant to contain the original string representation as well as the length, like TokenList and Uint?

**Comment by frewsxcv** [comment]
> I might be overlooking something here...

It is the case that I overlooked something here

**Comment by Ms2ger** [comment]
Make sure that `font.setAttribute("size", "01"); font.getAttribute("size") === "01"` passes.

**Comment by frewsxcv** [comment]
How do those changes look?

**Comment by frewsxcv** [comment]
Actually hold on, I forgot to incorporate the new `AttrValue` variant.

**Comment by frewsxcv** [comment]
Okay, ready to be reviewed now

**Comment by nox** [comment]
Reviewed 1 of 1 files at r2, 10 of 10 files at r3.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/attr.rs, line 30 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-32yJ3V7Pwnhtmr1sQ)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/attr.rs#L30)):</sup>
Just length, no? We will never have computed values in AttrValue, AFAICT.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 104 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-33LYk-Y5CRXSYBjjv)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L104)):</sup>
`value.as_length()`

---

<sup>**[components/script/dom/htmlfontelement.rs, line 119 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-33R4OKZpjeIZlTkav)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L119)):</sup>
Oh… I didn't think about this… I'm not so sure anymore about storing a Length anymore. :/ @jdm?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by nox** [comment]
-S-awaiting-review +S-awaiting-answer 

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 119 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-33R4OKZpjeIZlTkav)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L119)):</sup>
This seems like it does the right thing though, no? What are your concerns about how it's currently working?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/attr.rs, line 30 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-32yJ3V7Pwnhtmr1sQ)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/attr.rs#L30)):</sup>
https://github.com/servo/servo/pull/7683#issuecomment-142747924

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by nox** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 119 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-33R4OKZpjeIZlTkav)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L119)):</sup>
There is currently no attribute of which the variant can change depending on its value.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by Ms2ger** [comment]
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3TFDoYK7xe_1Sqhdz)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L28)):</sup>
No need for this.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 117 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3W4bEBekPV75KUMKs)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L117)):</sup>
A given (element type, attribute name) pair should always have the same enum variant.

---

<sup>**[components/style/values.rs, line 271 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3WECVOIS6Vgy95JvK)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/style/values.rs#L271)):</sup>
Is Length only used for font-size?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 117 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3W4bEBekPV75KUMKs)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L117)):</sup>
Can you think of any way to enforce this at compile time?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/style/values.rs, line 271 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3WECVOIS6Vgy95JvK)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/style/values.rs#L271)):</sup>
Thus far, yes.

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 117 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3W4bEBekPV75KUMKs)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L117)):</sup>
And also, how do you think I should handle this here?

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3TFDoYK7xe_1Sqhdz)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L28)):</sup>
Ah, you're right. That said, I'm not going to bother fixing it until the larger issue (multiple enum variants being used) is addressed

---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by nox** [comment]
@frewsxcv I had another suggestion in the beginning:

> I don't like the idea of storing DOMString here. You could at least store an Atom (they are only 7 small strings).

**Comment by Ms2ger** [comment]
We discussed this for a bit on IRC. Since we need to support attribute values like "01" being treated as "1" for the presentational hint, yet keep returning "01" from `getAttribute`, we'll go with either `Option<Length>` or `Option<enum { 1, 2, 3, 4, 5, 6, 7 }>`.

**Comment by frewsxcv** [comment]
Updated to use `Option<Length>` in the enum variant. Ready for a review again

**Comment by frewsxcv** [comment]
Actually hold off. I'm hitting a panic relating to accessing Element attributes on the style thread

```
 0:03.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) Full command: /Users/coreyf/Development/rust/servo/target/debug/servo --cpu --hard-fail -u Servo/wptrunner http://localhost:8000/_mozilla/mozilla/htmlfontelement_size_attribute.html --user-stylesheet /Users/coreyf/Development/rust/servo/resources/ahem.css
(pid:41960) "thread 'LayoutWorker worker 1/6' panicked at 'assertion failed: task_state::get().is_script()', /Users/coreyf/Development/rust/servo/components/script/dom/bindings/cell.rs:73"
 0:03.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "stack backtrace:"
 0:03.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   1:        0x10bcfc1c0 - sys::backtrace::tracing::imp::write::h88885f5c2d394b3flbt"
 0:03.36 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   2:        0x10bcfb323 - panicking::on_panic::h7a7974779bbfef81TZw"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   3:        0x10bcebaf2 - sys_common::unwind::begin_unwind_inner::h879cfd5a30161a9evas"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   4:        0x10a507e47 - sys_common::unwind::begin_unwind::h4358625394363150362"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   5:        0x10aa51fdc - dom::bindings::cell::DOMRefCell<T>::try_borrow::h4110685824665763101"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   6:        0x10aa51d2a - dom::bindings::cell::DOMRefCell<T>::borrow::h17323943642333816207"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   7:        0x10a5931c4 - dom::element::Element::get_attribute::hd1432472b4f05f01ImR"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   8:        0x10aae7371 - dom::htmlfontelement::HTMLFontElement::get_size::hae475f48a3e573ecU0X"
 0:03.37 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "   9:        0x10a4e868e - dom::element::Element.RawLayoutElementHelpers::synthesize_presentational_hints_for_legacy_attributes::h12023173141370083898"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  10:        0x10a4e7d21 - wrapper::LayoutElement<'le>.TElementAttributes::synthesize_presentational_hints_for_legacy_attributes::h6821834921749250198"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  11:        0x10a4e7cc1 - legacy::Stylist.PresentationalHintSynthesis::synthesize_presentational_hints_for_legacy_attributes::h5894885901084301425"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  12:        0x10a4d8a2e - selector_matching::Stylist::push_applicable_declarations::h15449980880287612884"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  13:        0x10a4c4bfb - css::matching::LayoutNode<'ln>.MatchMethods::match_node::hfec17b2f7f8a65e0kPA"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  14:        0x10a483597 - traversal::RecalcStyleForNode<'a>.PreorderDomTraversal::process::h8224edbbed6f075c05y"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  15:        0x10a482b8b - parallel::RecalcStyleForNode<'a>.ParallelPreorderDomTraversal::run_parallel::h0195bed7e2811b2ebcu"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  16:        0x10a483ee2 - parallel::recalc_style::hb8d11fd7b1615ac8Bcu"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  17:        0x10a392548 - workqueue::WorkerThread<QueueData, WorkData>::start::h608275555925366460"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  18:        0x10a391ccf - workqueue::WorkQueue<QueueData, WorkData>::new::closure.54686"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  19:        0x10a391b7a - task::spawn_named::closure.54679"
 0:03.38 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  20:        0x10a391ab9 - boxed::F.FnBox<A>::call_box::h16410021766752924616"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  21:        0x10a34dadb - boxed::Box<FnBox<A, Output $u3d$$u20$R$GT$$u2b$$u20$Send$u20$$u2b$$u20$$u27$a$GT$.FnOnce$LT$A$GT$::call_once::h6681894323494919348"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  22:        0x10a34d6bd - thread::Builder::spawn_inner::closure.52858"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  23:        0x10a34d669 - sys_common::unwind::try::try_fn::h6987218963915758806"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  24:        0x10bcfa9e8 - __rust_try"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  25:        0x10bcf7db0 - sys_common::unwind::try::inner_try::h30a7e5c7446717a0T6r"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  26:        0x10a34d5d3 - sys_common::unwind::try::h5411403924796213732"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  27:        0x10a34d44b - thread::Builder::spawn_inner::closure.52835"
 0:03.39 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  28:        0x10a34dcfc - boxed::F.FnBox<A>::call_box::h4746392297401831892"
 0:03.40 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  29:        0x10bcfce8d - sys::thread::Thread::new::thread_start::h6e92c6728040a269Mxw"
 0:03.40 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  30:     0x7fff90648059 - _pthread_body"
 0:03.40 PROCESS_OUTPUT: Thread-TestrunnerManager-1 (pid:41960) "  31:     0x7fff90647fd6 - _pthread_start"
```

**Comment by frewsxcv** [comment]
Just learned about `get_attr_for_layout`. This no longer panics and is reviewable again

**Comment by nox** [comment]
-S-awaiting-review +S-needs-code-changes 


---

Reviewed 5 of 5 files at r4.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/attr.rs, line 30 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-32yJ3V7Pwnhtmr1sQ)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/attr.rs#L30)):</sup>
Not sure if your link is about `SpecifiedLength` or not. I would rename that to `Length`. 


---

<sup>**[components/script/dom/attr.rs, line 111 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-Pknchbe2cc9_AIeyR)** ([raw file](https://github.com/servo/servo/blob/1822f91611b893c399c82002c5fa17d904ea67dd/components/script/dom/attr.rs#L111)):</sup>
`length.as_ref()`


---

<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3TFDoYK7xe_1Sqhdz)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L28)):</sup>
Mmmh @Ms2ger, why exactly didn't we store it just here instead of pushing it in Attr? Should the color and face fields go away too?


---

<sup>**[components/script/dom/htmlfontelement.rs, line 133 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-PlFPHGu2bkwMifikE)** ([raw file](https://github.com/servo/servo/blob/1822f91611b893c399c82002c5fa17d904ea67dd/components/script/dom/htmlfontelement.rs#L133)):</sup>
You can just `.clone()` here, no?


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
<sup>**[components/script/dom/htmlfontelement.rs, line 133 \[r4\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-PlFPHGu2bkwMifikE)** ([raw file](https://github.com/servo/servo/blob/1822f91611b893c399c82002c5fa17d904ea67dd/components/script/dom/htmlfontelement.rs#L133)):</sup>
TIL about `cloned` https://doc.rust-lang.org/stable/std/option/enum.Option.html#method.cloned


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by frewsxcv** [comment]
With exception to the comment @Ms2ger was mentioned in, all comments have been addressed

**Comment by frewsxcv** [comment]
Rebased

**Comment by nox** [comment]
Looks good, just waiting for @Ms2ger's answer now. Thanks for your extreme patience. :D


---

Reviewed 8 of 8 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, all commit checks successful.

---



Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by Ms2ger** [comment]
Answer to what?

**Comment by Ms2ger** [comment]
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

<sup>**[components/script/dom/htmlfontelement.rs, line 28 \[r3\]](https://reviewable.io:443/reviews/servo/servo/7683#-K-3TFDoYK7xe_1Sqhdz)** ([raw file](https://github.com/servo/servo/blob/9cd78762c6858535c43bec1b3c4b89bedb1cd1b4/components/script/dom/htmlfontelement.rs#L28)):</sup>
There are a number of places where we use this antipattern; they should all be removed.  For this PR, I'm happy to just stop adding new ones.


---


Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/servo/servo/7683)

**Comment by nox** [comment]
@bors-servo r+

Finally. :)

**Pull request merged**